src/dataflow/src/source/mod.rs

```diff
@@ -861,6 +867,117 @@ enum MessageProcessing {
     YieldedWithDelay,
 }
 
+#[derive(Clone)]
+/// A thread-safe shareable timestamp provider that allows consumers to pause ticking of the clock
+/// while they are in the middle of producing data for a particular timestamp.
+pub struct Timestamper {
+    inner: Arc<RwLock<Timestamp>>,
+}
+
+impl Timestamper {
+    fn new() -> Self {
+        let now = UNIX_EPOCH
+            .elapsed()
+            .unwrap()
+            .as_millis()
+            .try_into()
+            .unwrap();
+        Self {
+            inner: Arc::new(RwLock::new(now)),
+        }
+    }
+    /// Grab a lease for the current timestamp. This functions returns an opaque type that
+    /// dereferences to the timestamp and will automatically release the lease when dropped.
+    pub(crate) async fn lease(&self) -> impl Deref<Target = Timestamp> + '_ {
+        self.inner.read().await
+    }
+
+    /// Monotonically increases the current timestamp. This method waits for all current leases to
+    /// end before changing the value. Furthermore, while a tick is pending no more leases will be
+    /// given. This is due to the write-preferring behaviour of the tokio RwLock.
+    async fn tick(&self) -> Timestamp {
```

Comment:
I would feel a bit better if we tried harder to have each tick be strictly increasing I think. the code in `generate_next_timestamp` tries hard to be resilient to pathological clock resets which I think is nice

Reply:
Hm, what behaviour would you expect here? This is an async function so it could definitely sleep for a bit without causing problems to the caller, but this will block any readers that would like to emit new messages